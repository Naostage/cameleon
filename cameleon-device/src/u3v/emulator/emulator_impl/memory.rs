use cameleon_impl::memory::{memory, register_map};

const SBRM_ADDRESS: u64 = 0xffff;
const SIRM_ADDRESS: u64 = (SBRM::base() + SBRM::size()) as u64;

/// Offset | Value | Description.
///      0 |     1 | User Defined Name is supported.
///      1 |     0 | Access Privilege and Heartbeat are NOT supported.
///      2 |     0 | Message Channel is NOT supported.
///      3 |     1 | Timestampl is supported.
///    4-7 |  0000 | String Encoding (Ascii).
///      8 |     1 | Family Name is supported.
///      9 |     1 | SBRM is supported.
///     10 |     1 | Endianness Register is supported.
///     11 |     1 | Written Length Field is supported.
///     12 |     0 | Multi Event is currently NOT supported.
///     13 |     0 | Stacked Commands is NOT supported.
///     14 |     1 | Device Software Interface Version is supported.
///  15-63 |     0 | Reserved. All remained bits are set to 0.
const DEVICE_CAPABILITY: &[u8] = &[
    0b1001, 0b0000, 0b1111, 0b0010, 0b0000, 0b0000, 0b0000, 0b0000,
];

/// Offset | Value | Description.
///      0 |     0 | Heartbeat is not used.
///      1 |     0 | MultiEvent is not enabled.
///   2-63 |     0 | Reserved. All remained bits are set to 0.
const DEVICE_CONFIGURATION: &[u8] = &[
    0b0000, 0b0000, 0b0000, 0b0000, 0b0000, 0b0000, 0b0000, 0b0000,
];

/// Offset | Value | Description.
///      0 |     1 | SIRM is available.
///      1 |     1 | EIRM is available.
///      2 |     0 | IIDC is NOT available.
///   3-63 |     0 | Reserved. All remained bits are set to 0.
const U3V_CAPABILITY: &[u8] = &[
    0b1100, 0b0000, 0b0000, 0b0000, 0b0000, 0b0000, 0b0000, 0b0000,
];

#[memory]
pub(super) struct Memory {
    abrm: ABRM,
    sbrm: SBRM,
}

#[register_map(base = 0, endianness = LE)]
pub(super) enum ABRM {
    #[register(len = 2, access = RO, ty = u16)]
    GenCpVersionMinor = 1,

    #[register(len = 2, access = RO, ty = u16)]
    GenCpVersionMajor = 1,

    #[register(len = 64, access = RO, ty = String)]
    ManufacturerName = "cameleon",

    #[register(len = 64, access = RO, ty = String)]
    ModelName = "cameleon model",

    #[register(len = 64, access = RO, ty = String)]
    FamilyName = "cameleon family",

    #[register(len = 64, access = RO, ty = String)]
    DeviceVersion = "none",

    #[register(len = 64, access = RO, ty = String)]
    ManufacturerInfo = "none",

    #[register(len = 64, access = RO, ty = String)]
    SerialNumber,

    #[register(len = 64, access = RW, ty = String)]
    UserDefinedName,

    #[register(len = 8, access = RO, ty = Bytes)]
    DeviceCapability = DEVICE_CAPABILITY,

    #[register(len = 4, access = RO, ty = u32)]
    MaximumDeviceResponseTime = 500, // 500 ms.

    #[register(len = 8, access = RO, ty = u64)]
    ManifestTableAddress, // TODO: Define manifest table address,

    #[register(len = 8, access = RO, ty = u64)]
    SBRMAddress = SBRM_ADDRESS,

    #[register(len = 8, access = RW, ty = Bytes)]
    DeviceConfiguration = DEVICE_CONFIGURATION,

    #[register(len = 4, access = NA, ty = u32)]
    HeartbeatTimeout,

    #[register(len = 4, access = NA, ty = u32)]
    MessageChannelId,

    #[register(len = 8, access = RO, ty = u64)]
    Timestamp,

    #[register(len = 4, access = WO, ty = u32)]
    TimestampLatch,

    #[register(len = 8, access = RO, ty = u64)]
    TimestampIncrement = 1000, // Dummy value indicating device clock runs at 1MHZ.

    #[register(len = 4, access = NA, ty = Bytes)]
    AccessPrivilege,

    #[register(len = 4, access = RO, ty = Bytes)]
    ProtocolEndianness = &[0xFF, 0xFF, 0xFF, 0xFF], // Little endian.

    #[register(len = 4, access = NA, ty = Bytes)]
    ImplementationEndianness,

    #[register(len = 64, access = RO, ty = String)]
    DeviceSoftwareInterfaceVersion = "1.0.0",
}

#[register_map(base = SBRM_ADDRESS, endianness = LE)]
pub(super) enum SBRM {
    #[register(len = 2, access = RO, ty = u16)]
    U3VVersionMinor = 0,

    #[register(len = 2, access = RO, ty = u16)]
    U3VVersionMajor = 1,

    #[register(len = 8, access = RO, ty = Bytes)]
    U3VCapability = U3V_CAPABILITY,

    #[register(len = 4, access = RO, ty = u32)]
    MaximumCommandTransferLength = 1024,

    #[register(len = 4, access = RO, ty = u32)]
    MaximumAcknowledgeTransferLength = 1024,

    #[register(len = 4, access = RO, ty = u32)]
    NumberOfStrewamChannel = 1,

    #[register(len = 8, access = RO, ty = u64)]
    SirmAddress = SIRM_ADDRESS,

    #[register(len = 4, access = RO, ty = u32)]
    SirmLength, // TODO: Filled after SIRM register map is implemented.

    #[register(len = 8, access = RO, ty = u64)]
    EirmAddress, // TODO: Filled after SIRM register map is implemented.

    #[register(len = 4, access = RO, ty = u32)]
    EirmLength, // TODO: Filled after EIRM register map is implmeneted.

    #[register(len = 8, access = NA, ty = u64)]
    Iidc2Address,

    #[register(len=4, access = RO, ty=u32)]
    CurrentSpeed = 0b1000,
}
