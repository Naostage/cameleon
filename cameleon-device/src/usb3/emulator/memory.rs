use std::borrow::Cow;
use std::io::Write;

use super::{EmulatorError, Result};
use semver::Version;

pub struct Memory {
    raw: Vec<u8>,
}

impl Memory {
    pub(super) fn new(size: usize) -> Self {
        let raw = vec![0; size];
        let mut memory = Self { raw };
        memory
    }

    pub(super) fn from_raw(raw: Vec<u8>) -> Self {
        Self { raw }
    }

    pub(super) fn dump(&self) -> &[u8] {
        &self.raw
    }

    pub(super) fn read(&mut self, address: u64) -> Result<u8> {
        let address = address as usize;
        self.check_access(address)?;
        Ok(self.raw[address])
    }

    pub(super) fn write(&mut self, address: u64, data: u8) -> Result<()> {
        let address = address as usize;
        self.check_access(address)?;

        self.raw[address] = data;
        Ok(())
    }

    fn check_access(&self, address: usize) -> Result<()> {
        if self.raw.len() < address {
            Err(EmulatorError::MemoryAccessViolation)
        } else {
            Ok(())
        }
    }
}

const SBRM_ADDRESS: u64 = 0xffff;

/// offset | value | Description.
///      0 |     1 | User Defined Name is supported.
///      1 |     0 | Access Priviledge and Heartbeat are NOT supported.
///      2 |     0 | Message Channel is NOT supported.
///      3 |     1 | Timestampl is supported.
///      4 |  0000 | String Encoding (Ascii).
///      8 |     1 | Family Name is supported.
///      9 |     1 | SBRM is supported.
///     10 |     1 | Endianess Register is supported.
///     11 |     1 | Written Length Field is supported.
///     12 |     1 | Multi Event is supported.
///     13 |     1 | Stacked Commands is supported.
///     14 |     1 | Device Software Interface Version is supported.
///     15 |     0 | Reserved. All remained bits are set to 0.
const DEVICE_CAPABILITY: u64 = 0b111111100001001;

#[derive(Clone)]
pub struct ABRM {
    gen_cp_version: Version,
    manufacturer_name: Cow<'static, str>,
    model_name: Cow<'static, str>,
    family_name: Cow<'static, str>,
    device_version: Cow<'static, str>,
    manufacturer_info: Cow<'static, str>,
    serial_number: Cow<'static, str>,
    user_defined_name: Cow<'static, str>,
    device_capability: u64,
    maximum_device_response_time: u32,
    manifest_table_address: u64,
    sbrm_address: u64,
    device_configuration: u64,
    heartbeat_timeout: u32,
    message_channel_id: u32,
    timestamp: u64,
    timestamp_latch: u32,
    timestamp_increment: u64,
    access_privilege: u32,
    protocol_endianess: u32,
    implementation_endianess: u32,
    device_software_interface_version: &'static str,
}

impl ABRM {
    fn flush(&self, memory: impl Write) -> Result<()> {
        todo!();
    }
}

impl Default for ABRM {
    fn default() -> Self {
        Self {
            gen_cp_version: Version::new(1, 2, 0),
            manufacturer_name: "cameleon".into(),
            model_name: "cameleon model".into(),
            family_name: "cameleon family".into(),
            device_version: "none".into(),
            manufacturer_info: "".into(),
            serial_number: "0000".into(),
            user_defined_name: "none".into(),
            device_capability: DEVICE_CAPABILITY,
            maximum_device_response_time: 100,
            manifest_table_address: 0, // TODO: Define manifest address,
            sbrm_address: SBRM_ADDRESS,
            device_configuration: 0b00,
            heartbeat_timeout: 0,
            message_channel_id: 0,
            timestamp: 0,
            timestamp_latch: 0,
            timestamp_increment: 1000, // Dummy value indicating device clock runs at 1MHz.
            access_privilege: 0,
            protocol_endianess: 0xffff,       // Big endian.
            implementation_endianess: 0xffff, // Big endian.
            device_software_interface_version: "1.0.0",
        }
    }
}
