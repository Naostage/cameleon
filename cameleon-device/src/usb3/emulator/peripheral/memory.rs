use std::borrow::Cow;
use std::io::Write;
use std::ops::Range;

use byteorder::{WriteBytesExt, LE};
use semver::Version;

use crate::usb3::register_map::AccessRight;

use super::{EmulatorError, EmulatorResult};

pub(super) struct Memory {
    inner: Vec<u8>,
    protection: MemoryProtection,
}

impl Memory {
    pub(super) fn new(inner: Vec<u8>) -> Self {
        Self {
            protection: MemoryProtection::new(inner.len()),
            inner,
        }
    }

    pub(super) fn read_mem(&self, range: Range<usize>) -> EmulatorResult<&[u8]> {
        self.protection.verify_address_with_range(range.clone())?;

        if !self
            .protection
            .access_right_with_range(range.clone())
            .is_readable()
        {
            return Err(EmulatorError::AddressNotReadable);
        }

        Ok(&self.inner[range])
    }

    pub(super) fn write_mem(&mut self, address: usize, data: &[u8]) -> EmulatorResult<()> {
        let range = address..address + data.len();
        self.protection.verify_address_with_range(range.clone())?;
        if !self
            .protection
            .access_right_with_range(range.clone())
            .is_writable()
        {
            return Err(EmulatorError::AddressNotWritable);
        }

        self.inner[range].copy_from_slice(data);

        Ok(())
    }

    pub(super) fn set_access_right(
        &mut self,
        range: impl IntoIterator<Item = usize>,
        access_right: AccessRight,
    ) {
        self.protection
            .set_access_right_with_range(range, access_right)
    }
}

const SBRM_ADDRESS: u64 = 0xffff;

/// offset | value | Description.
///      0 |     1 | User Defined Name is supported.
///      1 |     0 | Access Privilege and Heartbeat are NOT supported.
///      2 |     0 | Message Channel is NOT supported.
///      3 |     1 | Timestampl is supported.
///    7-4 |  0000 | String Encoding (Ascii).
///      8 |     1 | Family Name is supported.
///      9 |     1 | SBRM is supported.
///     10 |     1 | Endianess Register is supported.
///     11 |     1 | Written Length Field is supported.
///     12 |     1 | Multi Event is supported.
///     13 |     1 | Stacked Commands is supported.
///     14 |     1 | Device Software Interface Version is supported.
///  63-15 |     0 | Reserved. All remained bits are set to 0.
const DEVICE_CAPABILITY: u64 = 0b111111100001001;

#[derive(Clone)]
pub struct ABRM {
    gen_cp_version: Version,
    manufacturer_name: Cow<'static, str>,
    model_name: Cow<'static, str>,
    family_name: Cow<'static, str>,
    device_version: Cow<'static, str>,
    manufacturer_info: Cow<'static, str>,
    serial_number: Cow<'static, str>,
    user_defined_name: Cow<'static, str>,
    device_capability: u64,
    maximum_device_response_time: u32,
    manifest_table_address: u64,
    sbrm_address: u64,
    device_configuration: u64,
    heartbeat_timeout: u32,
    message_channel_id: u32,
    timestamp: u64,
    timestamp_latch: u32,
    timestamp_increment: u64,
    access_privilege: u32,
    protocol_endianess: u32,
    implementation_endianess: u32,
    device_software_interface_version: &'static str,
}

macro_rules! string_setter {
    ($fn_name:ident, $prop:ident) => {
        pub fn $fn_name(&mut self, name: &str) -> EmulatorResult<()> {
            verify_str(name)?;
            self.$prop = name.to_owned().into();
            Ok(())
        }
    };
}

impl ABRM {
    string_setter!(set_model_name, model_name);
    string_setter!(set_family_name, family_name);
    string_setter!(set_device_version, device_version);
    string_setter!(set_manufacturer_info, manufacturer_info);
    string_setter!(set_serial_number, serial_number);
    string_setter!(set_user_defined_name, user_defined_name);

    pub(super) fn flush(&self, mut memory: impl Write) -> EmulatorResult<()> {
        memory.write_u16::<LE>(self.gen_cp_version.minor as u16)?;
        memory.write_u16::<LE>(self.gen_cp_version.major as u16)?;
        write_str(&mut memory, &self.model_name)?;
        write_str(&mut memory, &self.family_name)?;
        write_str(&mut memory, &self.device_version)?;
        write_str(&mut memory, &self.manufacturer_info)?;
        write_str(&mut memory, &self.serial_number)?;
        write_str(&mut memory, &self.user_defined_name)?;
        memory.write_u64::<LE>(self.device_capability)?;
        memory.write_u32::<LE>(self.maximum_device_response_time)?;
        memory.write_u64::<LE>(self.manifest_table_address)?;
        memory.write_u64::<LE>(self.sbrm_address)?;
        memory.write_u64::<LE>(self.device_configuration)?;
        memory.write_u32::<LE>(self.heartbeat_timeout)?;
        memory.write_u32::<LE>(self.message_channel_id)?;
        memory.write_u64::<LE>(self.timestamp)?;
        memory.write_u32::<LE>(self.timestamp_latch)?;
        memory.write_u64::<LE>(self.timestamp_increment)?;
        memory.write_u32::<LE>(self.access_privilege)?;
        memory.write_u32::<LE>(self.protocol_endianess)?;
        memory.write_u32::<LE>(self.implementation_endianess)?;
        write_str(&mut memory, self.device_software_interface_version)?;

        Ok(())
    }
}

impl Default for ABRM {
    fn default() -> Self {
        Self {
            gen_cp_version: Version::new(1, 3, 0),
            manufacturer_name: "cameleon".into(),
            model_name: "cameleon model".into(),
            family_name: "cameleon family".into(),
            device_version: "none".into(),
            manufacturer_info: "".into(),
            serial_number: "0000".into(),
            user_defined_name: "none".into(),
            device_capability: DEVICE_CAPABILITY,
            maximum_device_response_time: 100,
            manifest_table_address: 0, // TODO: Define manifest table address.
            sbrm_address: SBRM_ADDRESS,
            device_configuration: 0b00,
            heartbeat_timeout: 0,
            message_channel_id: 0,
            timestamp: 0,
            timestamp_latch: 0,
            timestamp_increment: 1000, // Dummy value indicating device clock runs at 1MHz.
            access_privilege: 0,
            protocol_endianess: 0xffff,       // Little endian.
            implementation_endianess: 0xffff, // Little endian.
            device_software_interface_version: "1.0.0",
        }
    }
}

fn verify_str(s: &str) -> EmulatorResult<()> {
    const STRING_LENGTH_LIMIT: usize = 64;

    if !s.is_ascii() {
        return Err(EmulatorError::InvalidString("string format is not ascii"));
    }

    // String in register must be 0 terminated.
    if s.as_bytes().len() > STRING_LENGTH_LIMIT - 1 {
        return Err(EmulatorError::InvalidString("string is too long."));
    }

    Ok(())
}

fn write_str(w: &mut impl Write, s: &str) -> EmulatorResult<()> {
    verify_str(s)?;
    w.write_all(s.as_bytes())?;
    Ok(w.write_u8(0)?) // 0 terminate.
}

/// Map each address to its access right.
/// Access right is represented by 2 bits and mapping is done in 2 steps described below.
/// 1. First step is calculating block corresponding to the address. 4 access rights is packed into a single block, thus the block
///    position is calculated by `address / 4`.
/// 2. Second step is extracting the access right from the block. The offset of the access right is calculated by
///    `address % 4 * 2`.
// TODO: Consider better representation.
struct MemoryProtection {
    inner: Vec<u8>,
    memory_size: usize,
}

impl MemoryProtection {
    fn new(memory_size: usize) -> Self {
        let len = if memory_size == 0 {
            0
        } else {
            (memory_size - 1) / 4 + 1
        };
        let inner = vec![0; len];
        Self { inner, memory_size }
    }

    fn set_access_right(&mut self, address: usize, access_right: AccessRight) {
        let block = &mut self.inner[address / 4];
        let offset = address % 4 * 2;
        let mask = !(0b11 << offset);
        *block = (*block & mask) | access_right.as_num() << offset;
    }

    fn access_right(&self, address: usize) -> AccessRight {
        let block = self.inner[address / 4];
        let offset = address % 4 * 2;
        AccessRight::from_num(block >> offset & 0b11)
    }

    fn access_right_with_range(&self, range: impl IntoIterator<Item = usize>) -> AccessRight {
        range
            .into_iter()
            .fold(AccessRight::RW, |acc, i| acc.meet(self.access_right(i)))
    }

    fn set_access_right_with_range(
        &mut self,
        range: impl IntoIterator<Item = usize>,
        access_right: AccessRight,
    ) {
        range
            .into_iter()
            .for_each(|i| self.set_access_right(i, access_right));
    }

    fn verify_address(&self, address: usize) -> EmulatorResult<()> {
        if self.memory_size <= address {
            Err(EmulatorError::InvalidAddress)
        } else {
            Ok(())
        }
    }

    fn verify_address_with_range(
        &self,
        range: impl IntoIterator<Item = usize>,
    ) -> EmulatorResult<()> {
        for i in range {
            self.verify_address(i)?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::AccessRight::*;
    use super::*;

    #[test]
    fn test_protection() {
        // [RO, RW, NA, WO, RO];
        let mut protection = MemoryProtection::new(5);
        protection.set_access_right(0, RO);
        protection.set_access_right(1, RW);
        protection.set_access_right(2, NA);
        protection.set_access_right(3, WO);
        protection.set_access_right(4, RO);

        assert_eq!(protection.inner.len(), 2);
        assert_eq!(protection.access_right(0), RO);
        assert_eq!(protection.access_right(1), RW);
        assert_eq!(protection.access_right(2), NA);
        assert_eq!(protection.access_right(3), WO);
        assert_eq!(protection.access_right(4), RO);

        assert_eq!(protection.access_right_with_range(0..2), RO);
        assert_eq!(protection.access_right_with_range(2..4), NA);
        assert_eq!(protection.access_right_with_range(3..5), NA);
    }

    #[test]
    fn test_verify_address() {
        let protection = MemoryProtection::new(5);
        assert!(protection.verify_address(0).is_ok());
        assert!(protection.verify_address(4).is_ok());
        assert!(protection.verify_address(5).is_err());
        assert!(protection.verify_address_with_range(2..5).is_ok());
        assert!(protection.verify_address_with_range(2..6).is_err());
    }

    #[test]
    fn test_read_write_memory() {
        let inner = vec![0, 0, 0, 0];
        let mut mem = Memory::new(inner);
        mem.set_access_right(0..4, RW);

        let data = &[1, 2, 3, 4];
        assert!(mem.write_mem(0, data).is_ok());
        assert_eq!(mem.read_mem(0..4).unwrap(), data);
    }
}
