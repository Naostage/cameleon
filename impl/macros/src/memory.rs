use proc_macro2::TokenStream;
use quote::quote;
use syn::{spanned::Spanned, Error, Result};

pub(super) fn expand(input: proc_macro::TokenStream) -> Result<proc_macro::TokenStream> {
    let memory_struct = MemoryStruct::parse(input)?;

    let expanded_struct = memory_struct.define_struct();
    let methods = memory_struct.impl_methods();
    let memory_trait = memory_struct.impl_memory_trait();

    Ok(proc_macro::TokenStream::from(quote! {
        #expanded_struct
        #memory_trait
        #methods
    }))
}

struct MemoryStruct {
    vis: syn::Visibility,
    ident: syn::Ident,
    attrs: Vec<syn::Attribute>,
    fragments: Vec<MemoryFragment>,
}

impl MemoryStruct {
    fn parse(input: proc_macro::TokenStream) -> Result<Self> {
        let input_struct: syn::ItemStruct = syn::parse(input)?;
        let span = input_struct.span();

        let ident = input_struct.ident;
        let attrs = input_struct.attrs;
        let vis = input_struct.vis;

        let mut fragments = vec![];
        match input_struct.fields {
            syn::Fields::Named(fields) => {
                for field in fields.named {
                    fragments.push(MemoryFragment::parse(field)?);
                }
            }
            other => {
                return Err(Error::new_spanned(other, "expected named field"));
            }
        }

        if fragments.is_empty() {
            return Err(Error::new(span, "at least one field required"));
        }

        Ok(Self {
            vis,
            ident,
            attrs,
            fragments,
        })
    }

    fn define_struct(&self) -> TokenStream {
        let ident = &self.ident;
        let vis = &self.vis;
        let attrs = &self.attrs;

        quote! {
            #(#attrs)*
            #vis struct #ident {
                raw: Vec<u8>,
                protection: cameleon_impl::memory::MemoryProtection,
                observers: std::vec::Vec<(std::ops::Range<usize>, std::boxed::Box<dyn cameleon_impl::memory::MemoryObserver>)>,
            }
        }
    }

    fn impl_methods(&self) -> TokenStream {
        let ident = &self.ident;
        let new = self.impl_new();
        let fragments_len = self.fragments.len();

        quote! {
            impl #ident {
                #new

                #[doc(hidden)]
                fn notify_all(&self, written_range: std::ops::Range<usize>) {

                    for (reg_range, observer) in &self.observers {

                        if written_range.start >= reg_range.end || written_range.end <= reg_range.start {
                            continue;
                        }
                        observer.update();
                    }
                }

                #[doc(hidden)]
                const fn calculate_memory_size(end_addresses: &[usize; #fragments_len]) -> usize {
                    let mut max = end_addresses[0];
                    let mut i = 0;
                    while i < #fragments_len {
                        if max < end_addresses[i] {
                            max = end_addresses[i];
                        }
                        i += 1;
                    }
                    max

                }
            }
        }
    }

    fn impl_memory_trait(&self) -> TokenStream {
        let ident = &self.ident;

        quote! {
            impl cameleon_impl::memory::prelude::MemoryRead for #ident {
                fn read_raw(&self, range: std::ops::Range<usize>) -> cameleon_impl::memory::MemoryResult<&[u8]> {
                    self.protection.verify_address_with_range(range.clone())?;
                    let access_right = self.protection.access_right_with_range(range.clone());
                    if !access_right.is_readable() {
                        return Err(cameleon_impl::memory::MemoryError::AddressNotReadable);
                    }

                    Ok(&self.raw[range])
                }

                fn read<T: cameleon_impl::memory::Register>(&self) -> cameleon_impl::memory::MemoryResult<T::Ty> {
                    T::read(&self.raw)
                }

                fn access_right<T: cameleon_impl::memory::Register>(&self) -> cameleon_impl::memory::AccessRight {
                    self.protection.access_right_with_range(T::range())
                }
            }

            impl cameleon_impl::memory::prelude::MemoryWrite for #ident {
                fn write_raw(&mut self, addr: usize, buf: &[u8]) -> cameleon_impl::memory::MemoryResult<()> {
                    let (start, end) = (addr, addr + buf.len());
                    let range = start..end;
                    self.protection.verify_address_with_range(range.clone())?;
                    let access_right = self.protection.access_right_with_range(range.clone());
                    if !access_right.is_writable() {
                        return Err(cameleon_impl::memory::MemoryError::AddressNotWritable);
                    }

                    self.raw[range].copy_from_slice(buf);
                    self.notify_all(start..end);

                    Ok(())
                }

                fn set_access_right<T: cameleon_impl::memory::Register>(&mut self, access_right: cameleon_impl::memory::AccessRight) {
                    self.protection.set_access_right_with_range(T::range(), access_right);
                }


                fn write<T: cameleon_impl::memory::Register>(&mut self, data: T::Ty) -> cameleon_impl::memory::MemoryResult<()>{
                    T::write(data, &mut self.raw)?;
                    self.notify_all(T::range());

                    Ok(())
                }

                fn register_observer<T, U>(
                    &mut self,
                    observer: U
                )
                    where T: cameleon_impl::memory::Register,
                          U: cameleon_impl::memory::MemoryObserver + 'static
                {
                    let reg_range = T::range();

                    self.observers.push((reg_range, Box::new(observer)));
                }

            }
        }
    }

    fn impl_new(&self) -> TokenStream {
        let vis = &self.vis;
        let memory_size = self.memory_size();

        let init_memory = self.fragments.iter().map(|f| {
            let ty = &f.ty;
            quote! {
                #ty::init_memory_protection(&mut protection);
                #ty::init_raw_memory(&mut raw);
            }
        });

        quote! {
            #vis fn new() -> Self {
                let mut raw = vec![0; #memory_size];
                let mut protection = cameleon_impl::memory::MemoryProtection::new(#memory_size);
                #(#init_memory)*
                Self {
                    raw,
                    protection,
                    observers: std::vec::Vec::new(),
                }
            }
        }
    }

    fn memory_size(&self) -> TokenStream {
        let end_addresses = self.fragments.iter().map(|f| {
            let size = f.size();
            let base = f.base();
            quote! {#size + #base}
        });

        quote! {
            Self::calculate_memory_size(&[#(#end_addresses),*])
        }
    }
}

struct MemoryFragment {
    ty: syn::Path,
}

impl MemoryFragment {
    fn parse(field: syn::Field) -> Result<Self> {
        let ty = match field.ty {
            syn::Type::Path(p) => p.path,
            other => return Err(Error::new_spanned(other, "expected type path")),
        };

        Ok(Self { ty })
    }

    fn base(&self) -> TokenStream {
        let ty = &self.ty;
        quote!(#ty::base())
    }

    fn size(&self) -> TokenStream {
        let ty = &self.ty;
        quote!(#ty::size())
    }
}
