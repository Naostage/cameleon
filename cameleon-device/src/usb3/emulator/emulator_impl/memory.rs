use cameleon_impl::memory::{memory, register};

const SBRM_ADDRESS: u64 = 0xffff;

// TODO: Multievent support.
/// Offset | Value | Description.
///      0 |     1 | User Defined Name is supported.
///      1 |     0 | Access Privilege and Heartbeat are NOT supported.
///      2 |     0 | Message Channel is NOT supported.
///      3 |     1 | Timestampl is supported.
///    7-4 |  0000 | String Encoding (Ascii).
///      8 |     1 | Family Name is supported.
///      9 |     1 | SBRM is supported.
///     10 |     1 | Endianness Register is supported.
///     11 |     1 | Written Length Field is supported.
///     12 |     0 | Multi Event is currentrly NOT supported.
///     13 |     1 | Stacked Commands is supported.
///     14 |     1 | Device Software Interface Version is supported.
///  63-15 |     0 | Reserved. All remained bits are set to 0.
const DEVICE_CAPABILITY: &[u8] = &[
    0b1001, 0b0000, 0b1111, 0b0110, 0b0000, 0b0000, 0b0000, 0b0000,
];

#[memory]
pub(super) struct Memory {
    abrm: ABRM,
}

#[register(base = 0, endianness = LE)]
pub(super) enum ABRM {
    #[entry(len = 2, access = RO, ty = u16)]
    GenCpVersionMinor = 1,

    #[entry(len = 2, access = RO, ty = u16)]
    GenCpVersionMajor = 1,

    #[entry(len = 64, access = RO, ty = String)]
    ManufacturerName = "cameleon",

    #[entry(len = 64, access = RO, ty = String)]
    ModelName = "cameleon model",

    #[entry(len = 64, access = RO, ty = String)]
    FamilyName = "cameleon family",

    #[entry(len = 64, access = RO, ty = String)]
    DeviceVersion = "none",

    #[entry(len = 64, access = RO, ty = String)]
    ManufacturerInfo = "none",

    #[entry(len = 64, access = RO, ty = String)]
    SerialNumber,

    #[entry(len = 64, access = RW, ty = String)]
    UserDefinedName,

    #[entry(len = 8, access = RO, ty = Bytes)]
    DeviceCapability = DEVICE_CAPABILITY,

    #[entry(len = 4, access = RO, ty = u32)]
    MaximumDeviceResponseTime = 100,

    #[entry(len = 8, access = RO, ty = u64)]
    ManifestTableAddress, // TODO: Define manifest table address,

    #[entry(len = 8, access = RO, ty = u64)]
    SBRMAddress = SBRM_ADDRESS,

    #[entry(len = 8, access = RO, ty = u64)]
    DeviceConfiguration,

    #[entry(len = 4, access = NA, ty = u32)]
    HeartbeatTimeout,

    #[entry(len = 4, access = NA, ty = u32)]
    MessageChannelId,

    #[entry(len = 8, access = RO, ty = u64)]
    Timestamp,

    #[entry(len = 4, access = WO, ty = u32)]
    TimestampLatch,

    #[entry(len = 8, access = RO, ty = u64)]
    TimestampIncrement = 1000, // Dummy value indicating device clock runs at 1MHZ.

    #[entry(len = 4, access = NA, ty = Bytes)]
    AccessPrivilege,

    #[entry(len = 4, access = RO, ty = Bytes)]
    ProtocolEndianness = &[0xFF, 0xFF, 0xFF, 0xFF], // Little endian.

    #[entry(len = 4, access = NA, ty = Bytes)]
    ImplementationEndianness,

    #[entry(len = 64, access = RO, ty = String)]
    DeviceSoftwareInterfaceVersion = "1.0.0",
}
