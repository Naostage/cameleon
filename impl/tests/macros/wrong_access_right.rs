/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

use cameleon_impl::memory::register_map;

#[register_map(base = 0, endianness = LE)]
pub enum ABRM {
    #[register(len = 2, access = Ro, ty = u16)]
    GenCpVersionMinor = 321,

    #[register(len = 8, access = RO, ty = u64)]
    SBRMAddress = SBRM_ADDRESS,
}

fn main() {}
