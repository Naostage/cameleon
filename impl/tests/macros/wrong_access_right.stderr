error: expected NA, RO, WO, or RW
 --> $DIR/wrong_access_right.rs:9:34
  |
9 |     #[register(len = 2, access = Ro, ty = u16)]
  |                                  ^^
